File: runners/go/algorithms/lsbx.go
package algorithms

import (
	"image"
)

type LSBX struct {
	Channel int
}

func (lsbx LSBX) Read(img image.Image, payload_data []byte) {
	g := img.Bounds()

	height := g.Dy()
	width := g.Dx()

	length := len(payload_data) * 8
	pos := 0

	channels := make([]uint32, 3)

	for i := 0; i < height; i++ {
		for j := 0; j < width; j++ {

			channels[0], channels[1], channels[2], _ = img.At(j, i).RGBA()

			if length <= 0 {
				return
			}

			payload_data[pos/8] = get_lsb(payload_data[pos/8], byte(channels[lsbx.Channel]))
			pos++
			length--
		}
	}
}

File: runners/csharp/csharp/Algorithms/lsbx.cs
using System.Drawing;

namespace Runner.Algorithms
{
    class LSBX : IAlgorithm
    {
        private int channel = 0;
        public LSBX(int channel) => this.channel = channel;


        void IAlgorithm.Read(Bitmap bm, byte[] payload_data)
        {
            int length = payload_data.Length * 8;
            int pos = 0;

            for (int i = 0; i < bm.Height; i++)
            {
                for (int j = 0; j < bm.Width; j++)
                {
                    Color pixel_color = bm.GetPixel(j, i);

                    byte[] channels = {
                        pixel_color.R,
                        pixel_color.G,
                        pixel_color.B
                    };

                    if (length <= 0)
                    {
                        return;
                    }

                    payload_data[pos / 8] = GetLsb(payload_data[pos / 8], channels[channel]);
                    pos++;
                    length--;
                }
            }
        }

        private static byte GetLsb(byte target, byte source)
        {
            return (byte)((target << 1) | (source & 1));
        }
    }
}
